apollo-router/src/plugins/file_uploads/mod.rs

```diff
@@ -0,0 +1,644 @@
+    let supergraph_result = req.extensions_mut().remove();
+    if let Some(supergraph_result) = supergraph_result {
+        let SubgraphHttpRequestExtensions {
+            multipart,
+            map_field,
+        } = supergraph_result;
+
+        let (mut request_parts, operations) = req.into_parts();
+        let form = MultipartFormData::new(operations, map_field, multipart);
+        request_parts
+            .headers
+            .insert(CONTENT_TYPE, form.content_type());
+        request_parts.headers.insert(APOLLO_REQUIRE_PREFLIGHT, TRUE);
+        return http::Request::from_parts(
+            request_parts,
+            hyper::Body::wrap_stream(form.into_stream().await),
+        );
+    }
+    req
+}
+
+#[derive(Clone)]
+struct MultipartRequest {
+    state: Arc<Mutex<MultipartRequestState>>,
+}
+
+struct MultipartRequestState {
+    multer: multer::Multipart<'static>,
+    limits: MultipartRequestLimits,
+    read_files_counter: usize,
+    file_sizes: Vec<usize>,
+    max_files_exceeded: bool,
+    max_files_size_exceeded: bool,
+}
+
+impl Drop for MultipartRequestState {
+    fn drop(&mut self) {
+        u64_counter!(
+            "apollo.router.operations.file_uploads",
+            "file uploads",
+            1,
+            "file_uploads.limits.max_file_size.exceeded" = self.max_files_size_exceeded,
+            "file_uploads.limits.max_files.exceeded" = self.max_files_exceeded
+        );
+
+        for file_size in &self.file_sizes {
+            u64_histogram!(
+                "apollo.router.operations.file_uploads.file_size",
+                "file upload sizes",
+                (*file_size) as u64
+            );
+        }
+        u64_histogram!(
+            "apollo.router.operations.file_uploads.files",
+            "number of files per request",
+            self.read_files_counter as u64
+        );
+    }
+}
+
+impl MultipartRequest {
+    fn new(request_body: hyper::Body, boundary: String, limits: MultipartRequestLimits) -> Self {
+        let multer = Multipart::with_constraints(
+            request_body,
+            boundary,
+            Constraints::new().size_limit(
+                SizeLimit::new().for_field("map", 10 * 1024), // hardcoded to 10kb
+            ),
+        );
+        Self {
+            state: Arc::new(Mutex::new(MultipartRequestState {
+                multer,
+                limits,
+                read_files_counter: 0,
+                file_sizes: Vec::new(),
+                max_files_exceeded: false,
+                max_files_size_exceeded: false,
+            })),
+        }
+    }
+
+    async fn operations_field(&mut self) -> UploadResult<multer::Field<'static>> {
+        Ok(self
+            .state
+            .lock()
+            .await
+            .multer
+            .next_field()
+            .await?
+            .filter(|field| field.name() == Some("operations"))
+            .ok_or_else(|| FileUploadError::MissingOperationsField)?)
```

Comment:
```suggestion
        self
            .state
            .lock()
            .await
            .multer
            .next_field()
            .await?
            .filter(|field| field.name() == Some("operations"))
            .ok_or_else(|| FileUploadError::MissingOperationsField)
```
if it works it's easier to read IMHO

Reply:
Clippy also got this: [c0a03c6](https://github.com/apollographql/router/pull/4584/commits/c0a03c6a3e392bc6274685f6e0ea3e05a5be2db4)